driver-core/src/main/com/mongodb/internal/operation/AsyncChangeStreamBatchCursor.java

```diff
@@ -102,22 +112,63 @@ public void close() {
 
     @Override
     public void setBatchSize(final int batchSize) {
-        wrapped.setBatchSize(batchSize);
+        getWrapped().setBatchSize(batchSize);
     }
 
     @Override
     public int getBatchSize() {
-        return wrapped.getBatchSize();
+        return getWrapped().getBatchSize();
     }
 
     @Override
     public boolean isClosed() {
-        return isClosed.get();
+        if (isClosed.get()) {
+            return true;
+        } else if (wrappedClosedItself()) {
+            close();
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private boolean wrappedClosedItself() {
+        AsyncAggregateResponseBatchCursor<RawBsonDocument> observedWrapped = wrapped.get();
+        return observedWrapped != null && observedWrapped.isClosed();
+    }
+
+    /**
+     * {@code null} is written to {@link #wrapped} before closing the wrapped object to maintain the following guarantee:
+     * if {@link #wrappedClosedItself()} observes a {@linkplain AsyncAggregateResponseBatchCursor#isClosed() closed} wrapped object,
+     * then it closed itself as opposed to being closed by {@code this}.
+     */
+    private void nullifyAndCloseWrapped() {
+        AsyncAggregateResponseBatchCursor<RawBsonDocument> observedWrapped = wrapped.getAndSet(null);
+        if (observedWrapped != null) {
+            observedWrapped.close();
+        }
+    }
+
+    /**
+     * This method guarantees that the {@code wrapped} argument is closed even if
+     * {@link #setWrappedOrCloseIt(AsyncAggregateResponseBatchCursor)} is called concurrently with or after (in the happens-before order)
+     * the method {@link #close()}.
+     */
+    private void setWrappedOrCloseIt(final AsyncAggregateResponseBatchCursor<RawBsonDocument> wrapped) {
```

Comment:
Nit: Use alternative variable name for the replacement wrapped. The differences between `wrapped` and `this.wrapped` can be made more explicit.

Reply:
I had the same momenentary confusion when reading.

Reply:
Renamed the parameter to `newValue`.